### PR TITLE
aruco: new feature testCharucoCornersCollinear() in charuco.hpp/cpp

### PR DESCRIPTION
**Pull request by amy-tabb**
When using charuco corners for calibration, specifically in the robotics context when full-pattern images of the charuco boards are not gauranteed, there are cases when the chaurco corners identified are collinear.

The corners (in my application) are usually axis aligned, but when these identified corners are axis aligned or just linear (some diagonal within the pattern), using these corners will result in errors via CV_ASSERT checks, or poor results b/c of the numerical issues of using collinear points for estimating calibration parameters.

There does not exist any functionality w/in the opencv_contrib module for checking whether the charuco corners are collinear, given a charuco board.  In `charuco.cpp`, there is function `static bool _arePointsEnoughForPoseEstimation(const vector< Point3f > &points)` called by  `estimatePoseCharucoBoard()`.  This function is local to `charuco.cpp`, though, and only checks for axis-aligned collinear charuco corners.

The new feature is a function in charuco.hpp/.cpp that checks for collinear charuco corners, `testCharucoCornersCollinear()`.

[mwe.zip](https://github.com/opencv/opencv_contrib/files/4539926/mwe.zip)

The MWE mwe.zip (with some input images) alters `calibrate_camera_charuco.cpp` to demonstrate situations in which calibration fails when there are collinear charuco corners.  In 3.4, the warning message  

```
** On entry to DLASCLS parameter number  4 had an illegal value
```

is displayed and the intrinisic camera calibration matrix contains nan values.

In 4.3/master, an exception is thrown:

```
terminate called after throwing an instance of 'cv::Exception'
  what():  OpenCV(4.3.0-dev) /home/atabb/git/opencv-charuco/charuco-bool-feature/opencv/modules/calib3d/src/calibration.cpp:1213: error: (-215:Assertion failed) fabs(sc) > DBL_EPSILON in function 'cvFindExtrinsicCameraParams2'
```

The image below illustrate one such image.  I did not have a diagonal case detected in my dataset, but created one to test.  

image_00018

image_00016_diag

Within the MWE, I show how I would use this new feature -- to choose whether to include (or not) detected charuco boards.  

```cpp
//			// TODO Uncomment this when testing the new feature.
//			bool collinear = testCharucoCornersCollinear(charucoboard, currentCharucoIds);
//
//			cout << "file " << vectorFiles[i] << " is collinear ? " << collinear << endl;
//
//			if (currentCharucoCorners.size() >= 6 && collinear== false){
//				{
//					allCharucoCorners.push_back(currentCharucoCorners);
//					allCharucoIds.push_back(currentCharucoIds);
//				}
//			}

			// TODO  comment this when testing the new feature
			if (currentCharucoCorners.size() >= 6){
				{
					allCharucoCorners.push_back(currentCharucoCorners);
					allCharucoIds.push_back(currentCharucoIds);
				}
			}

```

For this PR, I included everything in the PR versus putting the failure cases in a separate issue.  I am new to submitting PRs to OpenCV, let me know what works better.  I also used the 3.4 branch this time. 

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [X] I agree to contribute to the project under OpenCV (BSD) License.
- [X] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [X] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [X] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [X] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by amy-tabb** [comment]
The build failed b/c the parameter names didn't match those in the function.  I tried to fix that in charuco.hpp and committed the branch again.

File: modules/aruco/include/opencv2/aruco/charuco.hpp [hunk 1]
```diff
@@ -334,7 +334,18 @@ CV_EXPORTS void drawCharucoDiamond(const Ptr<Dictionary> &dictionary, Vec4i ids,
...
+ *
+ * @param _board layout of ChArUco board.
+ * @param _charucoIds list of identifiers for each corner in charucoCorners per frame.
+ * @return bool value, 1 (true) if detected corners form a line, 0 (false) if they do not.
+      solvePnP, calibration functions will fail if the corners are collinear (true).
+ *
+ * The number of ids in charucoIDs should be <= the number of chessboard corners in the board.  This functions checks whether the charuco corners are on a straight line (returns true, if so), or not (false).  Axis parallel, as well as diagonal and other straight lines detected.  Degenerate cases: for number of charucoIDs <= 2, the function returns true.
+ */
+CV_EXPORTS_W bool testCharucoCornersCollinear(const Ptr<CharucoBoard> &_board,
+                                              InputArray _charucoIds);
```
Comment (alalek) [inline modules/aruco/include/opencv2/aruco/charuco.hpp:348]: > _board

You can use parameter names without underscore symbol in headers (no need to synchronize with .cpp)

File: modules/aruco/src/charuco.cpp [hunk 2]
```diff
@@ -938,5 +938,62 @@ void drawDetectedDiamonds(InputOutputArray _image, InputArrayOfArrays _corners,
...
+    Vec<double, 3> testLine = point0.cross(point1);
+
+    Vec<double, 3> testPoint(0, 0, 1);
+
+    double divisor = sqrt(testLine[0]*testLine[0] + testLine[1]*testLine[1]);
+
+    CV_Assert( divisor != 0);
+
+     // normalize the line with normal
+     testLine /= divisor;
```
Comment (alalek) [inline modules/aruco/src/charuco.cpp:977]: Please fix indentation

File: modules/aruco/src/charuco.cpp [hunk 3]
```diff
@@ -938,5 +938,62 @@ void drawDetectedDiamonds(InputOutputArray _image, InputArrayOfArrays _corners,
...
+      solvePnP will fail if the corners are collinear (true).
+  * Check that the set of charuco markers in _charucoIds does not identify a straight line on
+    the charuco board.  Axis parallel, as well as diagonal and other straight lines detected.
+  */
+bool testCharucoCornersCollinear(const Ptr<CharucoBoard> &_board, InputArray _charucoIds){
+
+    unsigned int nCharucoCorners = (unsigned int)_charucoIds.getMat().total();
+
+    // only test if there are 3 or more corners
+    if (nCharucoCorners > 2){
```
Comment (alalek) [inline modules/aruco/src/charuco.cpp:955]:
```
if (nCharucoCorners <= 2)
    return true;
```

File: modules/aruco/include/opencv2/aruco/charuco.hpp (hunk 1 above)
Comment (amy-tabb) [reply modules/aruco/include/opencv2/aruco/charuco.hpp:348]: OK

File: modules/aruco/src/charuco.cpp (hunk 2 above)
Comment (amy-tabb) [reply modules/aruco/src/charuco.cpp]: Fixed in latest commit.

File: modules/aruco/src/charuco.cpp (hunk 3 above)
Comment (amy-tabb) [reply modules/aruco/src/charuco.cpp]: I switched the order of the test in this function -- which I think is what you're suggesting here.

**Comment by amy-tabb** [comment]
Re: `add simple test for testCharucoCornersCollinear()` can you refer me to some ref materials, and/or give me some idea of what you have in mind? What would be tested?  I read the performance tests page from https://github.com/opencv/opencv/wiki/How_to_contribute and it didn't seem quite relevant.

**Comment by alalek** [comment]
Consider adding test into [this file](https://github.com/opencv/opencv_contrib/blob/3.4.10/modules/aruco/test/test_charucodetection.cpp).

No need to put/read image somewhere, just hard-code corner values in test code [directly](https://github.com/opencv/opencv/blob/c722625f280258f5c865002899bf0dc2ebff1b2b/modules/calib3d/test/test_cameracalibration.cpp#L1992-L2012) (you can dump them from your example pictures above).

```
TEST(Charuco, testCharucoCornersCollinear_true)
{
    ... create CharucoBoard and fill hard-coded collinear corners ...
    ... with identifiers ...
    bool result = cv::testCharucoCornersCollinear(board, charucoIds);
    EXPECT_TRUE(result);
}

TEST(Charuco, testCharucoCornersCollinear_false)
{
    ... create CharucoBoard and fill hard-coded corners ...
    ... with identifiers ...
    bool result = cv::testCharucoCornersCollinear(board, charucoIds);
    EXPECT_FALSE(result);
}
```

**Comment by amy-tabb** [comment]
Ok -- thanks for the explanation, which I understood.  I'll give this a try another day this week.

**Comment by amy-tabb** [comment]
I created the test and ran it successfully on my system.  Before the build system went down for maintenance, it indicated the test code wasn't in compliance w/ C++98, so I tried to fix that on the last commit.
